src/classify/classifier/nearest_centroid.cpp

```diff
@@ -75,22 +112,23 @@ class_label nearest_centroid::classify(doc_id d_id)
     return best_label;
 }
 
+template <class ForwardIterator>
 double nearest_centroid::cosine_sim(
-    const std::vector<std::pair<term_id, double>>& doc,
-    const std::unordered_map<term_id, double>& centroid)
+    ForwardIterator begin, ForwardIterator end,
+    const std::unordered_map<term_id, double>& centroid) const
```

Comment:
maybe I should switch this to `sparse_vector` so we can get a faster `cosine_sim`...